pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -375,6 +401,215 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
+        quay_url = constants.QUAY_RELEASE_REPO_URL
+        # Anything under this directory will be sync'd to the mirror
+        BASE_TO_MIRROR_DIR = f"{working_dir}/to_mirror/openshift-v4"
+        shutil.rmtree(f"{BASE_TO_MIRROR_DIR}/{arch}", ignore_errors=True)
+
+        # From the newly built release, extract the client tools into the workspace following the directory structure
+        # we expect to publish to mirror
+        CLIENT_MIRROR_DIR = f"{BASE_TO_MIRROR_DIR}/{arch}/clients/{client_type}/{release_name}"
+        os.makedirs(CLIENT_MIRROR_DIR)
+
+        # extract release clients tools
+        extract_release_client_tools(f"{quay_url}:{from_release_tag}", f"--to={CLIENT_MIRROR_DIR}", None)
+
+        # Get cli installer operator-registory pull-spec from the release
+        for tarball in ["cli", "installer", "operator-registry"]:
+            image_stat, cli_pull_spec = get_release_image_pullspec(f"{quay_url}:{from_release_tag}", tarball)
+            if image_stat == 0:  # image exists
+                _, image_info = get_release_image_info_from_pullspec(cli_pull_spec)
+                # Retrieve the commit from image info
+                commit = image_info["config"]["config"]["Labels"]["io.openshift.build.commit.id"]
+                source_url = image_info["config"]["config"]["Labels"]["io.openshift.build.source-location"]
+                source_name = source_url.split("/")[-1]
+                source_name = constants.MIRROR_CLIENTS[source_name]
+                # URL to download the tarball a specific commit
+                response = requests.get(f"{source_url}/archive/{commit}.tar.gz", stream=True)
+                if response.ok:
+                    with open(f"{CLIENT_MIRROR_DIR}/{source_name}-src-{from_release_tag}.tar.gz", "wb") as f:
+                        f.write(response.raw.read())
+                    # calc shasum
+                    with open(f"{CLIENT_MIRROR_DIR}/{source_name}-src-{from_release_tag}.tar.gz", 'rb') as f:
+                        shasum = hashlib.sha256(f.read()).hexdigest()
+                    # write shasum to sha256sum.txt
+                    with open(f"{CLIENT_MIRROR_DIR}/sha256sum.txt", 'a') as f:
+                        f.write(f"{shasum}  {source_name}-src-{from_release_tag}.tar.gz\n")
+
+        if arch == 'x86_64':
+            # oc image  extract requires an empty destination directory. So do this before extracting tools.
+            # oc adm release extract --tools does not require an empty directory.
+            image_stat, oc_mirror_pullspec = get_release_image_pullspec(f"{quay_url}:{from_release_tag}", "oc-mirror")
+            if image_stat == 0:  # image exist
+                # extract image to workdir, if failed it will raise error in function
+                extract_release_binary(oc_mirror_pullspec, [f"--path=/usr/bin/oc-mirror:{CLIENT_MIRROR_DIR}"])
+                current_path = os.getcwd()
+                os.chdir(CLIENT_MIRROR_DIR)
```

Comment:
Avoid `os.chdir` if possible. This could lead to race condition if we run those functions with multithread or coroutine in the future.

Reply:
if we don't chdir it will make symlink content in tarball incorrect like the issue https://redhat-internal.slack.com/archives/CJARLA942/p1683773060618819

Reply:
This is not true. symlinks have nothing to do with current directory or chdir call. What you need to do is setting the symlink link target to a relative path instead of full physical path.